common/src/main/java/org/keycloak/common/util/StringSerialization.java

```diff
@@ -0,0 +1,116 @@
+package org.keycloak.common.util;
+
+import java.net.URI;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Utilities to serialize objects to string. Type safety is not guaranteed here and is responsibility of the caller.
+ * @author hmlnarik
+ */
+public class StringSerialization {
+
+    // Since there is still need to support JDK 7, we have to work without functional interfaces
+    private static enum DeSerializer {
+        OBJECT {
+            @Override public String serialize(Object o)   { return o.toString(); };
+            @Override public Object deserialize(String s) { return s; };
+        },
+        URI {
+            @Override public String serialize(Object o)   { return o.toString(); };
+            @Override public Object deserialize(String s) { return java.net.URI.create(s); };
+        },
+        ;
+
+        /** Serialize value which is guaranteed to be non-null */
+        public abstract String serialize(Object o);
+        public abstract Object deserialize(String s);
+    }
+
+    private static final Map<Class<?>, DeSerializer> WELL_KNOWN_SERIALIZERS = new LinkedHashMap<>();
+    private static final Map<Class<?>, DeSerializer> WELL_KNOWN_DESERIALIZERS = new LinkedHashMap<>();
+    private static final String SEPARATOR = ";";
+    private static final Pattern ESCAPE_PATTERN = Pattern.compile(SEPARATOR);
+    private static final Pattern UNESCAPE_PATTERN = Pattern.compile(SEPARATOR + SEPARATOR);
+    private static final Pattern VALUE_PATTERN = Pattern.compile("([NV])" +
+      "(" +
+        "(?:[^" + SEPARATOR + "]|" + SEPARATOR + SEPARATOR + ")*?" +
+      ")($|" + SEPARATOR + "(?!" + SEPARATOR + "))",
+      Pattern.DOTALL
+    );
+
+    static {
+        WELL_KNOWN_SERIALIZERS.put(URI.class, DeSerializer.URI);
```

Comment:
It seems redundant to use two maps, isn't better to have a single one `WELL_KNOWN_SERIALIZAERS` ?

Reply:
You're absolutely right, this was omission from JDK 8 -> JDK 7 transition.